pkg/operator/latencyprofilecontroller/controller.go

```diff
@@ -0,0 +1,188 @@
+
+	controlplanev1 "github.com/openshift/api/kubecontrolplane/v1"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/operatorclient"
+)
+
+const (
+	apiServerConfigMapName = "config"
+	apiServerConfigMapKey  = "config.yaml"
+
+	DefaultNotReadyTolerationSecondsArgument    = "default-not-ready-toleration-seconds"
+	DefaultUnreachableTolerationSecondsArgument = "default-unreachable-toleration-seconds"
+)
+
+// LatencyProfileController periodically lists the config/v1/node object
+// and fetches the worker latency profile applied on the cluster which is used to
+// updates the status of the config node object. The status updates reflect the
+// state of kube-api-server(s) running on control plane node(s) and their
+// observed config for default-not-ready-toleration-seconds and
+// default-unreachable-toleration-seconds match the applied arguments.
+
+type LatencyProfileController struct {
+	operatorClient   v1helpers.StaticPodOperatorClient
+	configClient     configv1.ConfigV1Interface
+	configMapClient  corev1client.ConfigMapsGetter
+	configNodeLister listerv1.NodeLister
+}
+
+func NewLatencyProfileController(
+	operatorClient v1helpers.StaticPodOperatorClient,
+	configClient configv1.ConfigV1Interface,
+	nodeInformer configv1informers.NodeInformer,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	kubeClient kubernetes.Interface,
+	eventRecorder events.Recorder,
+) factory.Controller {
+
+	ret := &LatencyProfileController{
+		operatorClient:   operatorClient,
+		configClient:     configClient,
+		configMapClient:  v1helpers.CachedConfigMapGetter(kubeClient.CoreV1(), kubeInformersForNamespaces),
+		configNodeLister: nodeInformer.Lister(),
+	}
+
+	return factory.New().WithInformers(
+		// this is for our general configuration input and our status output in case another actor changes it
+		operatorClient.Informer(),
+
+		// We use nodeInformer for observing current worker latency profile
+		nodeInformer.Informer(),
+
+		// for configmaps of operator client target namespace
+		kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Informer(),
+	).ResyncEvery(time.Minute).WithSync(ret.sync).WithSyncDegradedOnError(operatorClient).ToController(
+		"LatencyProfileController",
+		eventRecorder.WithComponentSuffix("latency-profile-controller"),
+	)
+}
+
+func (c *LatencyProfileController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	_, operatorStatus, _, err := c.operatorClient.GetStaticPodOperatorState()
+	if err != nil {
+		return err
+	}
+
+	// Collect the unique set of revisions of the apiserver nodes
+	revisionMap := map[int32]struct{}{}
+	uniqueRevisions := []int32{}
+	for _, nodeStatus := range operatorStatus.NodeStatuses {
+		revision := nodeStatus.CurrentRevision
+		if _, ok := revisionMap[revision]; !ok {
+			revisionMap[revision] = struct{}{}
+			uniqueRevisions = append(uniqueRevisions, revision)
+		}
+	}
+
+	// Collect the current latency profile
```

Comment:
I'd move it before collecting the revisions (line 86). Then I'd read the profile and check if there is work to do. Does it make sense?

Reply:
Yeah, did that too.